android/src/main/java/io/ably/flutter/plugin/push/PushBackgroundIsolateRunner.java

```diff
@@ -0,0 +1,101 @@
+    this.broadcastReceiver = receiver;
+    this.remoteMessage = message;
+    flutterEngine = new FlutterEngine(context, null);
+    DartExecutor executor = flutterEngine.getDartExecutor();
+    backgroundMethodChannel = new MethodChannel(executor.getBinaryMessenger(), "io.ably.flutter.plugin.background", new StandardMethodCodec(new AblyMessageCodec()));
+    backgroundMethodChannel.setMethodCallHandler(this);
+    // Get and launch the users app isolate manually:
+    executor.executeDartEntrypoint(DartExecutor.DartEntrypoint.createDefault());
+    // Even though lifecycle parameter is @NonNull, the implementation `FlutterEngineConnectionRegistry`
+    // does not use it, because it is a bug in the API design. See https://github.com/flutter/flutter/issues/90316
+    flutterEngine.getBroadcastReceiverControlSurface().attachToBroadcastReceiver(receiver, null);
+  }
+
+  /**
+   * This method is called when the main app is running and the user sets the background handler.
+   *
+   * @param backgroundMessageHandlerHandle
+   */
+  public static void setBackgroundMessageHandler(Context context, Long backgroundMessageHandlerHandle) {
+    SharedPreferences preferences = context.getApplicationContext().getSharedPreferences(SHARED_PREFERENCES_KEY, MODE_PRIVATE);
+    preferences.edit().putLong(BACKGROUND_MESSAGE_HANDLE_KEY, backgroundMessageHandlerHandle).apply();
+  }
+
+  @Override
+  public void onMethodCall(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
+    if (call.method.equals(pushSetOnBackgroundMessage)) {
+      // This signals that the manually spawned app is ready to receive a message to handle.
+      // We ask the user to set the background message handler early on.
+      backgroundMethodChannel.invokeMethod(pushOnBackgroundMessage, remoteMessage, new MethodChannel.Result() {
+            @Override
+            public void success(@Nullable Object result) {
+              finish();
+            }
+
+            @Override
+            public void error(String errorCode, @Nullable String errorMessage, @Nullable Object errorDetails) {
+              Log.e(TAG, String.format("pushOnBackgroundMessage method call from Java to Dart returned error.\n " +
+                  "errorCode: %s\n" +
+                  "errorMessage: %s\n" +
+                  "errorDetails: %s\n", errorCode, errorMessage, errorDetails));
+              finish();
+            }
+
+            @Override
+            public void notImplemented() {
+              Log.e(TAG, "Method: pushOnBackgroundMessage did not receive message on dart side. " +
+                  "Either the binding has not been initialized or the method call handler was not registered to receive this method.");
+              finish();
+            }
+          }
+      );
+    } else {
+      result.notImplemented();
+    }
+  }
+
+  private void finish() {
+    assert flutterEngine != null;
```

Comment:
I've used it because it is an assumption in that the FlutterEngine was running when you call finish, and if the assumption breaks that is a SDK developer mistake. If I just fail early, I'm telling users of this function that it's okay to mutate `flutterEngine`, close it early, or just set it to `null`. I don't want this, because `finish()` is the only thing that should nullify `flutterEngine`. TBH, it doesn't matter because finish also calls `FirebaseMessagingReceiver.finish();`, which kills the entire Receiver. Most importantly, what happens if `flutterEngine` was set to `null`, but was not `destroy`'d? The assert enforces a contract, a user should not be mutating `flutterEngine`.

It also only happens in debug mode: https://stackoverflow.com/a/6176529/7365866 and https://stackoverflow.com/a/2758645/7365866

Reply:
I don't know much about the mechanism of Flutter engine, but why do you allow users to mutate this? Should this not be a part of some sort of lifecycle?

Reply:
26a1862b2a7c44a75574d84f5a326fd3dce5a1db

Reply:
Thanks, both. 😄